zrml/prediction-markets/src/migrations.rs

```diff
@@ -157,7 +172,148 @@ impl<T: Config> OnRuntimeUpgrade for MigrateMarketPoolsBeforeOpen<T> {
                 );
             }
         }
+        Ok(())
+    }
+}
+
+impl<T: Config> OnRuntimeUpgrade for CleanUpStorageForResolvedOrClosedMarkets<T> {
+    fn on_runtime_upgrade() -> Weight
+    where
+        T: Config,
+    {
+        let mut total_weight = T::DbWeight::get().reads(1);
+
+        if StorageVersion::get::<Pallet<T>>()
+            != PREDICTION_MARKETS_REQUIRED_STORAGE_VERSION_FOR_CLEANUP_STORAGE_FOR_RESOLVED_MARKETS
+        {
+            log::info!("Skipping storage cleanup; prediction-markets already up to date");
+            return total_weight;
+        }
+
+        total_weight = total_weight.saturating_add(T::DbWeight::get().reads(2));
+        log::info!("Starting storage cleanup of CleanUpStorageForResolvedOrClosedMarkets");
+
+        let dispute_period = T::DisputePeriod::get();
+        let current_block: T::BlockNumber = <frame_system::Pallet<T>>::block_number();
+        let last_dp_end_block = current_block.saturating_sub(dispute_period);
```

Comment:
As far as I can tell, the runtime upgrades trigger _before_ the block's `on_initialize` is called:  https://github.com/paritytech/substrate/blob/8e600ee0515122667cd091d632104708d343ca34/frame/executive/src/lib.rs#L292-L319

This means that in terms of disputing, the chain state lags one block behind during upgrades.

```suggestion
        let last_dp_end_block = current_block.saturating_sub(dispute_period).saturating_sub(1);
```

Reply:
Then should it be  `let last_dp_end_block = current_block.saturating_sub(dispute_period).saturating_add(1);` ?

Reply:
I might have made a mistake. My reasoning is this: If we're at the start of block `x`, then the last resolve we did was on block `x - 1`, so all markets that have been reported/disputed on block `x - 1 - DisputePeriod` have already been resolved. So these can be cleared. But we haven't resolved markets that have been reported/disputed on block `x - DisputePeriod`, because we havn't yet run `on_initialize` on block `x`. So `last_dp_end_block` must be `x - 1`.

Right...?

Reply:
but from resolution_manager() we can see that for block X it will be resolved 
`let block = now.saturating_sub(dispute_period);` and from PR it will remove storage.

Reply:
Sorry, I'm not sure what you mean by "[...] and from PR it will remove storage." But just to emphasize my point: If you set `last_dp_end_block` to `x - 1 - DisputePeriod`, then on block `x`, the update will be executed before `on_initialize`, and the migration will clear the cache of all markets reported/disputed at block `x - 1 - DisputePeriod` or before. Suppose we have a market that was disputed on block `x - DisputePeriod`. This market's cache will _not_ be deleted. After the update, `on_initialize` will resolve the market and then remove it's cache.

Contrary, if you let `last_dp_end_block = x - DisputePeriod` and there's a market that was reported/disputed on block `x - DisputePeriod`, it will still be `Reported`/`Disputed` at the start of block `x`, and the market's cache will be deleted by the migration. When `on_initialize` is called for block `x`, it will no longer find this market in the cache and the market will never be resolved.

Reply:
makes sense now!